services/venice-controller/src/main/java/com/linkedin/venice/pushmonitor/PushMonitorUtils.java

```diff
@@ -0,0 +1,112 @@
+        : ExecutionStatus.COMPLETED;
+    ExecutionStatus middleStatus = incrementalPushVersion.isPresent()
+        ? ExecutionStatus.START_OF_INCREMENTAL_PUSH_RECEIVED
+        : ExecutionStatus.END_OF_PUSH_RECEIVED;
+    Optional<String> erroredInstance = Optional.empty();
+    String storeName = Version.parseStoreFromKafkaTopicName(topicName);
+    int completedPartitions = 0;
+    int totalInstanceCount = 0;
+    int liveInstanceCount = 0;
+    Set<Integer> incompletePartition = new HashSet<>();
+    for (int partitionId = 0; partitionId < partitionCount; partitionId++) {
+      Map<CharSequence, Integer> instances = reader.getPartitionStatus(
+          storeName,
+          Version.parseVersionFromVersionTopicName(topicName),
+          partitionId,
+          incrementalPushVersion);
+      boolean allInstancesCompleted = true;
+      totalInstanceCount += instances.size();
+      for (Map.Entry<CharSequence, Integer> entry: instances.entrySet()) {
+        ExecutionStatus status = ExecutionStatus.fromInt(entry.getValue());
+        boolean isInstanceAlive = reader.isInstanceAlive(storeName, entry.getKey().toString());
+        if (!isInstanceAlive) {
+          continue;
+        }
+        // We only compute status based on live instances.
+        liveInstanceCount++;
+        if (status == middleStatus) {
+          if (allInstancesCompleted) {
+            allInstancesCompleted = false;
+          }
+        } else if (status != completeStatus) {
+          if (allInstancesCompleted || allMiddleStatusReceived) {
+            allInstancesCompleted = false;
+            allMiddleStatusReceived = false;
+            if (status == ExecutionStatus.ERROR) {
+              erroredInstance = Optional.of(entry.getKey().toString());
+              break;
+            }
+          }
+        }
+      }
+      if (allInstancesCompleted) {
+        completedPartitions++;
+      } else {
+        incompletePartition.add(partitionId);
+      }
+    }
+    String statusDetail = null;
+    String details = "";
+    if (completedPartitions > 0) {
+      details += completedPartitions + "/" + partitionCount + " partitions completed in" + totalInstanceCount
+          + " Da Vinci instances.";
+    }
+    if (erroredInstance.isPresent()) {
+      details += "Found a failed instance in Da Vinci: " + erroredInstance + ". live instances: " + liveInstanceCount
+          + " total instances : " + totalInstanceCount;
+    }
+    int incompleteSize = incompletePartition.size();
+    if (incompleteSize > 0 && incompleteSize <= 5) {
+      details += ". Following partitions still not complete " + incompletePartition + ". live instances: "
+          + liveInstanceCount + " total instances : " + totalInstanceCount;
+    }
+    if (details.length() != 0) {
+      statusDetail = details;
+    }
+    if (completedPartitions == partitionCount) {
+      return new ExecutionStatusWithDetails(completeStatus, statusDetail);
+    } else if (allMiddleStatusReceived) {
```

Comment:
nit: we can simplify it a bit by removing the `else`, i.e.
`else if (allMiddleStatusReceived) {` ->  `if (allMiddleStatusReceived) {`